File: ago-dic-2020/AldahirBriones/Parcial1/MergeSort.java
public class MergeSort
{
    public void sort(int arr[], int left, int right)
    {
        if(left < right)
        {
            //Encuentra el punto medio del vector.
            int middle = (left + right) / 2;
      
            //Divide la primera y segunda mitad (llamada recursiva).
            sort(arr, left, middle);
            sort(arr, middle+1, right);

            //Une las mitades.
            merge(arr, left, middle, right);
        }
    }

    public void merge(int arr[], int left, int middle, int right) {
        //Encuentra el tamaño de los sub-vectores para unirlos.
        int n1 = middle - left + 1;
        int n2 = right - middle;

        //Vectores temporales.
        int leftArray[] = new int [n1];
        int rightArray[] = new int [n2];

        //Copia los datos a los arrays temporales.
        for (int i=0; i < n1; i++) 
        {
            leftArray[i] = arr[left+i];
        }
        for (int j=0; j < n2; j++) 
        {
            rightArray[j] = arr[middle + j + 1];
        }
        /* Une los vectorestemporales. */

        //Índices inicial del primer y segundo sub-vector.
        int i = 0, j = 0;

        //Índice inicial del sub-vector arr[].
        int k = left;

        //Ordenamiento.
        while (i < n1 && j < n2) 
        {
            if (leftArray[i] <= rightArray[j]) {
                arr[k] = leftArray[i];
                i++;
            } else {
                arr[k] = rightArray[j];
                j++;
            }
            k++;
        }//Fin del while.

        /* Si quedan elementos por ordenar */
        //Copiar los elementos restantes de leftArray[].
        while (i < n1) 
        {
            arr[k] = leftArray[i];
            i++;
            k++;
        }
        //Copiar los elementos restantes de rightArray[].
        while (j < n2) 
        {
            arr[k] = rightArray[j];
            j++;
            k++;
        }
    }

    public void printArray(int arr[]) 
    {
        int n = arr.length;
        for (int i=0; i<n; ++i) {
            System.out.println(arr[i] + " ");
        }
        System.out.println();
    }
    
    public static void main(String[] args) 
    {

        MergeSort mergeSort = new MergeSort();
        int arr [] = {5,26,12,6,1,4,7};
        int n = arr.length;

        System.out.println("Array original:");
        for (int value : arr) 
        {
            System.out.print(value + " ");
        }

        System.out.println();

        System.out.println("Array ordenado:");
        mergeSort.sort(arr,0,n-1);
        mergeSort.printArray(arr);
    }
}

File: ene-jun-2021/material/algoritmos/mergesort.py
def Merge(sorted_left_array, sorted_right_array):
    arr_resultado = []
    while len(sorted_left_array) > 0 and len(sorted_right_array) > 0:
        if sorted_left_array[0] > sorted_right_array[0]:
            arr_resultado.append(sorted_right_array[0])
            sorted_right_array.pop(0)
        else:
            arr_resultado.append(sorted_left_array[0])
            sorted_left_array.pop(0)
    
    while len(sorted_left_array) > 0:
        arr_resultado.append(sorted_left_array[0])
        sorted_left_array.pop(0)

    while len(sorted_right_array) > 0:
        arr_resultado.append(sorted_right_array[0])
        sorted_right_array.pop(0)
    
    return arr_resultado

def mergeSort(arr):
    if len(arr) == 1:
        return arr

    middle = len(arr) // 2
    left_array = arr[:middle]
    right_array = arr[middle:]

    sorted_left_array = mergeSort(left_array)
    sorted_right_array = mergeSort(right_array)

    return Merge(sorted_left_array, sorted_right_array)



arr = [1, 4,56,564,2,7,321,12,56,568,32,23,567,568,2,21,753,6]
arrNombre = ["mipeli1", "peli2", "peli3"]

print(arr)
arr2 = mergeSort(arr, arrNombre)
print(arr2)